### ISSUE TITLE
Cannot add account 

### ISSUE THREAD

**Issue by winster** [microsoft/azuredatastudio#15342]
Azure data studio, version: 1.28.0
OS: macOS Catalina and BigSur
I tried with and without http_proxy & https_proxy. Error remains same.

Console log:
`Studio.app/Contents/Resources/app/node_modules.asar/rxjs/observable/ForkJoinObservable.js.map: 
console.ts:137 [Extension Host] https://login.microsoftonline.com/common/oauth2/token - [{"grant_type":"authorization_code","code":"0.AxxxxxA","client_id":"a69788ssssss55","code_verifier":"_9OIrx50ddddddw","redirect_uri":"https://vscode-redirect.azurewebsites.net/","resource":"https://management.core.windows.net/"},"<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">\n<html>\n\n\n<head>\n  <meta content=\"text/html; charset=UTF-8\" http-equiv=\"Content-Type\">\n  <meta http-equiv=\"X-UA-Compatible\" content=\"IE=7\" />\n  <title>McAfee Web Gateway - Notification</title>\n  <script src=\"/mwg-internal/de5fs23hu73ds/files/javascript/sw.js\" type=\"text/javascript\" ></script>\n  <link rel=\"stylesheet\" href=\"/mwg-internal/de5fs23hu73ds/files/default/stylesheet.css\" />\n</head>\n\n\n<body onload=\"swOnLoad();\">\n  <table class='bodyTable'>\n    <tr>\n      <td class='bodyData' background='/mwg-internal/de5fs23hu73ds/files/default/img/bg_body.gif'>\n\n<table class='logoTable'>\n  <tr>\n    <td class='logoData'>\n      <a href='http://www.wal-mart.com'>\n        <img src='/mwg-internal/de5fs23hu73ds/files/default/img/walmart_logo.png'></a>\n    </td>\n  </tr>\n</table>\n\n\n\n\n<table class='titleTable' background='/mwg-internal/de5fs23hu73ds/files/default/img/bg_navbar.jpg'>\n  <tr>\n    <td class='titleData'>\n      Handshake failed\n    </td>\n  </tr>\n</table>\n\n\n\n<table class=\"contentTable\">\n  <tr>\n    <td class=\"contentData\">\n      The SSL handshake could not be performed.\n    </td>\n  </tr>\n</table>\n\n\n\n<table class=\"infoTable\">\n  <tr>\n    <td class=\"infoData\">\n      <b>Host: </b>login.microsoftonline.com<br />\n      <b>Reason: </b>Can&apos;t initialize server context:handshakefailed:server state 1:state 9:Application response 500 handshakefailed\n    </td>\n  </tr>\n</table>\n\n\n\n\n\n\n\n<table class='footTable'>\n  <tr>\n    <td class='helpDeskData' background='/mwg-internal/de5fs23hu73ds/files/default/img/bg_navbar.jpg'>\n      For assistance, please contact the Help Desk.\n    </td>\n  </tr>\n  <tr>\n    <td class='footData'>\n      Generated <span id=\"time\">2021-05-06 10:07:14</span> by mcweb-e6 on 9080<br>Rule Name = DevApps_Whitelist<br>40.126.29.13<br>172.28.109.146\n    </td>\n  </tr>\n</table>\n\n      </td>\n    </tr>\n  </table>\n</body>\n\n\n</html>"]
notificationsAlerts.ts:40 {}
onDidChangeNotification @ notificationsAlerts.ts:36
(anonymous) @ breadcrumbsModel.ts:294
fire @ event.ts:535
addNotification @ notifications.ts:204
notify @ notificationService.ts:73
(anonymous) @ mainThreadLogService.ts:16
ZoneAwarePromise @ /private/var/folders/v4/myl1yrg92sj_7nx2bfc833280000gn/T/AppTranslocation/7DDD9E47-C77B-493A-96F8-C7CD7BEEF558/d/Azure Data Studio.app/Contents/Resources/app/node_modules.asar/zone.js/dist/zone.js:891
_showMessage @ mainThreadLogService.ts:16
$showMessage @ mainThreadLogService.ts:16
_doInvokeHandler @ rpcProtocol.ts:351
_invokeHandler @ rpcProtocol.ts:351
_receiveRequest @ rpcProtocol.ts:296
_receiveOneMessage @ rpcProtocol.ts:237
(anonymous) @ rpcProtocol.ts:104
fire @ event.ts:535
fire @ ipc.net.ts:431
_receiveMessage @ ipc.net.ts:772
(anonymous) @ ipc.net.ts:628
fire @ event.ts:535
acceptChunk @ ipc.net.ts:241
(anonymous) @ ipc.net.ts:171
t @ ipc.net.ts:920
emit @ events.js:223
addChunk @ _stream_readable.js:309
readableAddChunk @ _stream_readable.js:290
Readable.push @ _stream_readable.js:224
onStreamRead @ internal/stream_base_commons.js:181
console.ts:137 [Extension Host] Error: No access token returned from Microsoft OAuth - 
t.log @ console.ts:141
$logExtensionHostMessage @ mainThreadConsole.ts:24
_doInvokeHandler @ rpcProtocol.ts:351
_invokeHandler @ rpcProtocol.ts:351
_receiveRequest @ rpcProtocol.ts:296
_receiveOneMessage @ rpcProtocol.ts:237
(anonymous) @ rpcProtocol.ts:104
fire @ event.ts:535
fire @ ipc.net.ts:431
_receiveMessage @ ipc.net.ts:772
(anonymous) @ ipc.net.ts:628
fire @ event.ts:535
acceptChunk @ ipc.net.ts:241
(anonymous) @ ipc.net.ts:171
t @ ipc.net.ts:920
emit @ events.js:223
addChunk @ _stream_readable.js:309
readableAddChunk @ _stream_readable.js:290
Readable.push @ _stream_readable.js:224
onStreamRead @ internal/stream_base_commons.js:181
log.ts:197   ERR Cannot read property 'accountType' of undefined: TypeError: Cannot read property 'accountType' of undefined
    at e.renderElement (file:///private/var/folders/v4/myl1yrg92sj_7nx2bfc833280000gn/T/AppTranslocation/7DDD9E47-C77B-493A-96F8-C7CD7BEEF558/d/Azure Data Studio.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:2536:51)
    at e.renderElement (file:///private/var/folders/v4/myl1yrg92sj_7nx2bfc833280000gn/T/AppTranslocation/`

**Comment by winster** [comment]
If I open this link https://management.core.windows.net/ directly in browser, a 2 way handshake is initiated which pops up a few certificates and none of the certificate responded other than
`<Error xmlns="http://schemas.microsoft.com/windowsazure" xmlns:i="http://www.w3.org/2001/XMLSchema-instance">
<Code>ForbiddenError</Code>
<Message>The server failed to authenticate the request. Verify that the certificate is valid and is associated with this subscription.</Message>
</Error>`

**Comment by cssuh** [comment]
Hi @winster, can you try using device code to see if you can authenticate using this method?

**Comment by jseabold** [comment]
Experiencing this same issue. Unable to login to azure through ADS on my mac. VS Code extensions are able to login.

**Comment by jseabold** [comment]
Doing some more investigating these all seem similar #15560 #17356 #13950. I'm also getting to the localhost redirect and it just sits with pending. I don't see anything listening on the port that it's redirecting to. I'm also behind a corporate proxy, which I suspect could be part of the issue. I've tried bypassing the proxy with no luck yet.

**Comment by Charles-Gagnon** [comment]
@jseabold Have you tried using device code auth? That should hopefully at least let you add your account if you're having firewall issues with the code grant method. 

![image](https://user-images.githubusercontent.com/28519865/141193734-c8ca7448-d244-4ee9-801d-d1ecadbed7ef.png)

Check device code and uncheck code grant.

**Comment by jseabold** [comment]
Saw that in another issue. Device code doesn't work. It looks like it does nothing and then throws the error above in the dev console.

I _did_ just get auth code to work. It was the proxy. I unset all of http_proxy, etc., rebooted because I don't know how the app gets the environment variables, made sure proxy was unset in the preferences, then I actually got off my corporate network and it worked. It's definitely some kind of proxy setting that's not allowing 127.0.0.1:xxxx to talk back to the app. I had 127.0.0.1 and loopback in my NO_PROXY environment variable, so not entirely sure what the issue was.

**Comment by jseabold** [comment]
Apparently, it needs to do the handshake every time you start the app, which I guess makes sense. This makes my proxy situation a little untenable. I'll see if there's something I can do locally, but if there's any chance this is something that's solvable on the app side that would be helpful. This is and the mssql plugin for vs code or the only apps where I've seen this behavior. 

If I unset my http_proxy, etc. and restart it works. If I unset them locally and/or try to curl with the `--noproxy '*'` it also just hangs. I've tried using `NO_PROXY`, and I've tried setting don't use the proxy in my system settings without any luck. Funnily enough, if I change the callback code the request gets through and returns an error which makes me think it's something other than the proxy. If I switch back to the correct callback code then it just hangs again.

<img width="538" alt="Screen Shot 2021-11-11 at 2 17 45 PM" src="https://user-images.githubusercontent.com/296164/141363351-fcb2a95c-7784-4c54-9cfa-95bcd922ca1f.png">

**Comment by preethamrajk** [comment]
@jseabold 

Could you confirm if your corporate proxy is ZScaler?

For a short duration, we had a different proxy and there was no issue then. But this started happening only after we switched to ZScaler.

**Comment by jseabold** [comment]
I'm not sure what proxy we use. I'm slightly less convinced this is a proxy error and not something in the application. I have to refresh my token, have all of my proxy variables unset and removed in system and application preferences, and I'm seeing the same behavior again. 

If I curl with an incorrect code, I get a response from ADS. If I curl with the correct code from the browser, it just hangs. Pretty frustrating.

**Comment by xanterx** [comment]
Thanks, @jseabold 

> I unset all of http_proxy

Unsetting all the https and http proxy environment variable worked for me as well.

- `HTTPS_PROXY`
- `HTTP_PROXY`

**Comment by cssuh** [comment]
Closing as we have made a number of proxy-related fixes, (most notably changing the redirect-uri to localhost and changing to use httpclient) and there has been no activity on this issue in a few months. Please feel free to open a new issue if you are still having problems.
